lib/src/arithmetic/partial_product_generator.dart

```diff
@@ -0,0 +1,260 @@
+// Copyright (C) 2024 Intel Corporation
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// partial_product_generator.dart
+// Partial Product matrix generation from Booth recoded multiplicand
+//
+// 2024 May 15
+// Author: Desmond Kirkpatrick <desmond.a.kirkpatrick@intel.com>
+
+import 'dart:math';
+
+import 'package:meta/meta.dart';
+import 'package:rohd/rohd.dart';
+import 'package:rohd_hcl/rohd_hcl.dart';
+
+/// A [PartialProductGenerator] class that generates a set of partial products.
+///  Essentially a set of
+/// shifted rows of [Logic] addends generated by Booth recoding and
+/// manipulated by sign extension, before being compressed
+class PartialProductGenerator {
+  /// Get the shift increment between neighboring product rows
+  int get shift => selector.shift;
+
+  /// The actual shift in each row. This value will be modified by the
+  /// sign extension routine used when folding in a sign bit from another
+  /// row
+  final rowShift = <int>[];
+
+  /// rows of partial products
+  int get rows => partialProducts.length;
+
+  /// The multiplicand term
+  Logic get multiplicand => selector.multiplicand;
+
+  /// The multiplier term
+  Logic get multiplier => encoder.multiplier;
+
+  /// Partial Products output. Generated by selector and extended by sign
+  /// extension routines
+  late final List<List<Logic>> partialProducts;
+
+  /// Encoder for the full multiply operand
+  late final MultiplierEncoder encoder;
+
+  /// Selector for the multiplicand which uses the encoder to index into
+  /// multiples of the multiplicand and generate partial products
+  late final MultiplicandSelector selector;
+
+  /// Operands are signed
+  final bool signed;
+
+  /// Used to avoid sign extending more than once
+  bool isSignExtended = false;
+
+  /// Construct a [PartialProductGenerator] -- the partial product matrix
+  PartialProductGenerator(
+      Logic multiplicand, Logic multiplier, RadixEncoder radixEncoder,
+      {required this.signed}) {
+    encoder = MultiplierEncoder(multiplier, radixEncoder, signed: signed);
+    selector =
+        MultiplicandSelector(radixEncoder.radix, multiplicand, signed: signed);
+
+    if (multiplicand.width < selector.shift) {
+      throw RohdHclException('multiplicand width must be greater than '
+          '${selector.shift}');
+    }
+    if (multiplier.width < (selector.shift + (signed ? 1 : 0))) {
+      throw RohdHclException('multiplier width must be greater than '
+          '${selector.shift + (signed ? 1 : 0)}');
+    }
+    _build();
+    signExtend();
+  }
+
+  /// Perform sign extension (defined in child classes)
+  @protected
+  void signExtend() {}
```

Comment:
Shoudl this be an `abstract` class with no definition for this function?  Maybe the "none" version should be a baseclass that overrides this function and does nothing?

Reply:
I wonder if I should call this class PartialProductGeneratorBase.

Then my default PartialProductGenerator can really be the PartialProductGenerator_WithCompactRectSIgnExtension.  I don't want to lose the other cases because when we are testing a new kind of arithmetic, we may have to drop back to, say, BruteForce, just to get something working and then fix whatever is broken in the fancier, optimized sign extension.

Reply:
Sure, that sounds good